js/hw_2_js_functions_state/src/game_of_snake/game_of_snake.js

```diff
@@ -1,3 +1,409 @@
+  if(state.getWin || state.getLose  || state.pause) return state
+  
+  if(state.timer === 0){
+      return {
+          ...state,
+          getLose: true
+      }
+  }
+
+  return {
+      ...state,
+      timer: state.timer - 100,
+      timeGone: state.timeGone + 100
+  }
+})
+
+
+engine.addSignalReducer('up', state => {
+  if(state.getWin || state.getLose || state.pause) return state
+
+
+  if(state.direction === 'up' || state.direction === 'down'){
+    return state;
+  }
+  return {
+      ...state,
+      direction: 'up',
+  };
+})
+
+engine.addSignalReducer('down', state => {
+  if(state.getWin || state.getLose || state.pause) return state
+
+
+  if(state.direction === 'down'  || state.direction === 'up'){
+    return state;
+  }
+  return {
+      ...state,
+      direction: 'down',
+  };
+})
+
+engine.addSignalReducer('left', state => {
+  if(state.getWin || state.getLose || state.pause) return state
+
+
+  if(state.direction === 'left'  || state.direction === 'right'){
+    return state;
+  }
+  return {
+      ...state,
+      direction: 'left',
+  };
+})
+
+engine.addSignalReducer('right', state => {
+  if(state.getWin || state.getLose || state.pause) return state
+
+
+  if(state.direction === 'right'  || state.direction === 'left'){
+    return state;
+  }
+  return {
+      ...state,
+      direction: 'right',
+  };
+})
+
+engine.addSignalReducer('snakeMove', state => {
+  const prevBodyPosition = state.snakeBody;
+  let prevHeadPosition = prevBodyPosition[0];
+  let newSnakeHead = state.snakeHead
+  let snakeLength = state.snakeBody.length
+
+  switch (state.direction) {
+    case 'down':
+      newSnakeHead = [prevHeadPosition[0], prevHeadPosition[1] + 1]
+      break;
+    
+    case 'up':
+      newSnakeHead = [prevHeadPosition[0], prevHeadPosition[1] - 1]
+      break;
+
+    case 'left':
+      newSnakeHead = [prevHeadPosition[0] - 1, prevHeadPosition[1]]
+      break;
+    
+    case 'right':
+      newSnakeHead = [prevHeadPosition[0] + 1, prevHeadPosition[1]]
+      break;
+  
+    default:
+      break;
+  }
+
+
+  if(isHitBorder(newSnakeHead) || isHitSelf(newSnakeHead, state.snakeBody)) {
+    return {
+      ...state,
+      getLose: true
+    }
+  } else if(state.pause) {
+    return state
+  }
+
+  let newData = []
+  let newFoodPosition = state.food
+  let newScore = state.score
+  let newTimeGone = state.timeGone
+
+  if (isFoodEaten(newSnakeHead, state.food)) {
+    newData = [
+      newSnakeHead,
+      ...state.snakeBody
+    ]
+
+    newFoodPosition = generateFood(state.snakeBody)
+    newScore = state.score + 1
+    newTimeGone = 0
+  }else if(state.timeGone % 15000 === 0) {
+    newData = [
+      newSnakeHead,
+      ...state.snakeBody.slice(0, snakeLength - 1)
+    ]
+
+    newFoodPosition = generateFood(state.snakeBody)
+    newTimeGone = 0
+  } else {
+    newData = [
+      newSnakeHead,
+      ...state.snakeBody.slice(0, snakeLength - 1)
+    ]
+  }
+
+  return {
```

Comment:
So, I like that you managed to put most of business logic in the reducer, however:
- avoid using `let` - it is hard to maintain code when things get re-assigned all the time
- move more logic in pure functions that live outside (good example - how you handled `isFoodEaten` and other utilities)
- avoid `switch` - but if you need it, move it elsewhere (a separate pure function is a good place) - it makes code harder to read
- make your reducer as thin as possible by extracting as much logic as possible into pure functions